073-decouple-ca-management-from-end-entity-certificate-management.md

```diff
@@ -0,0 +1,363 @@
+
+ZooKeeperReconciler and KafkaReconciler classes:
+1. Generate the certificates the Kafka and ZooKeeper nodes will present to clients (other ZK/Kafka nodes, the Cluster Operator, Kafka producer and consumers etc) and store them in the Kubernetes Secrets listed above, with the cluster-ca-cert-generation (and clients-ca-cert-generation for Kafka) annotation set to the current value passed in with the Cluster CA (and Clients CA).
+2. Create/update the StrimziPodSet used for the nodes in the cluster and add the following annotations:
+   1. cluster-ca-cert-generation set to the current values passed in with the Cluster CA.
+   2. cluster-ca-key-generation set to the current values passed in with the Cluster CA.
+   3. server-cert-hash set to the full SHA1-hash of the certificate that was generated.
+
+EntityOperatorReconciler, CruiseControlReconciler and KafkaExporterReconciler generate the certificates they will use to authenticate with Kafka and ZooKeeper and store them in the Kubernetes Secrets listed above, with the cluster-ca-cert-generation set to the current value passed in with the Cluster CA.
+
+### Example flows
+
+When a new cluster is first created the flow is:
+
+1. Either the user or the CaReconciler creates the Cluster CA certificate and key and stores them in the <CLUSTER_NAME>-cluster-ca-cert and <CLUSTER_NAME>-cluster-ca Kubernetes Secrets respectively.
+2. The CaReconciler generates the Cluster Operator certificate and stores it in the <CLUSTER_NAME>-cluster-operator-certs Kubernetes Secret.
+3. The component reconcilers (KafkaReconciler, ZooKeeperReconciler, EntityOperatorReconciler etc) generate their corresponding EE certificates (signed by the newly generated Cluster CA certificate), store them in Kubernetes Secrets and annotate the Kubernetes Secrets and Pods with the cert and key generations, and (for ZooKeeper and Kafka) the certificate hash.
+
+#### Fig 3: New cluster create flow
+
+![New cluster create flow](./images/073-cert-new-existing.png)
+
+> Fig 3: The numbers in the diagram match the numbers in the full description of the steps.
+> The diagram omits the use of annotations.
+
+When the Cluster CA private key is updated (either because the user has requested for it to be regenerated (while Strimzi is managing the CA) or the user has incremented the generation to indicate they've manually updated the Kubernetes Secret):
+1. If Strimzi is managing the Cluster CA, the CaReconciler regenerates the Cluster CAs certificate and key. The old certificate and key are kept in the respective Kubernetes Secret and renamed to ca-YYYY-MM-DDTHH-MM-SSZ.crt.
+2. The CaReconciler skips updating the <CLUSTER_NAME>-cluster-operator-certs Kubernetes Secret (since the components still only trust the old CA certificate).
+3. The CaReconciler rolls the ZooKeeper and Kafka nodes to trust the new CA certificates.
```

Comment:
At this point, I think both the old and new CAs are trusted at the same time. Is this correct?

Reply:
Correct